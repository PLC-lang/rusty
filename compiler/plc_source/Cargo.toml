[package]
name = "plc_source"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
encoding_rs.workspace = true
encoding_rs_io.workspace = true

[lib]
name = "plc_source"
path = "./src/lib.rs"
