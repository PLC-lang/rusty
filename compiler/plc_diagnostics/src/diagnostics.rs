use std::fmt::Display;

use serde::{Deserialize, Serialize};

use plc_ast::ast::AstNode;
use plc_source::{
    source_location::{SourceLocation, SourceLocationFactory},
    SourceCode,
};

pub mod diagnostics_registry;

pub const INTERNAL_LLVM_ERROR: &str = "internal llvm codegen error";

/// a diagnostics severity
#[derive(Default, Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Ignore,
    #[default]
    Info,
    Warning,
    Error,
}

/// The `Diagnostics` struct describes an issue encountered during compile time.
/// The issue is defined by an `error_code` and had a defined `severity`
/// Diagnostic severity can be overridden when being reported.
#[derive(Debug)]
pub struct Diagnostic {
    /// The Description of the error being reported.
    message: String,
    /// Primary location where the diagnostic occurred
    primary_location: SourceLocation,
    /// Seconday locations relevant to the diagnostics
    secondary_locations: Option<Vec<SourceLocation>>,
    /// Error code for reference in the documentation
    error_code: &'static str,
    /// Children of the current diagnostic
    sub_diagnostics: Vec<Diagnostic>,
    /// If the diagnostic is caused by an error, this field contains the original error
    internal_error: Option<anyhow::Error>,
}

impl std::error::Error for Diagnostic {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.internal_error.as_ref().and_then(|it| it.source())
    }
}

impl From<std::io::Error> for Diagnostic {
    fn from(value: std::io::Error) -> Self {
        Diagnostic::new(value.to_string()).with_error_code("E002").with_internal_error(value.into())
    }
}

/// Builder for Diagnostics
impl Diagnostic {
    pub fn new(message: impl Into<String>) -> Self {
        Diagnostic {
            message: message.into(),
            primary_location: SourceLocation::undefined(),
            secondary_locations: Default::default(),
            error_code: "E001", //Default error if none specified
            sub_diagnostics: Default::default(),
            internal_error: Default::default(),
        }
    }

    pub fn with_location<T>(mut self, location: T) -> Self
    where
        T: Into<SourceLocation>,
    {
        self.primary_location = location.into();
        self
    }

    pub fn with_secondary_location<T>(mut self, location: T) -> Self
    where
        T: Into<SourceLocation>,
    {
        self.secondary_locations.get_or_insert_with(Default::default).push(location.into());
        self
    }

    pub fn with_secondary_locations(mut self, locations: Vec<SourceLocation>) -> Self {
        self.secondary_locations.get_or_insert_with(Default::default).extend(locations);
        self
    }

    pub fn with_error_code(mut self, error_code: &'static str) -> Self {
        self.error_code = error_code;
        self
    }

    pub fn with_sub_diagnostic(mut self, diagnostic: Diagnostic) -> Self {
        self.sub_diagnostics.push(diagnostic);
        self
    }

    pub fn with_sub_diagnostics(mut self, diagnostics: Vec<Diagnostic>) -> Self {
        self.sub_diagnostics.extend(diagnostics);
        self
    }

    pub fn with_internal_error(mut self, error: anyhow::Error) -> Self {
        self.internal_error = Some(error);
        self
    }

    pub fn from_serde_error(error: serde_json::Error, source: &SourceCode) -> Self {
        let factory = SourceLocationFactory::for_source(source);
        let line = error.line();
        let column = error.column();

        // remove line, column from message
        let message = error.to_string();
        let message = if let Some(pos) = message.find("at line") {
            message.chars().take(pos).collect()
        } else {
            message
        };
        let range = factory.create_range_to_end_of_line(line, column);
        Diagnostic::new(message).with_error_code("E088").with_location(range)
    }
}

impl PartialEq for Diagnostic {
    fn eq(&self, other: &Self) -> bool {
        self.error_code == other.error_code
            && self.message == other.message
            && self.primary_location == other.primary_location
            && self.secondary_locations == other.secondary_locations
            && self.sub_diagnostics == other.sub_diagnostics
    }
}

impl Eq for Diagnostic {}

impl Display for Diagnostic {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.get_message())?;
        let location = self.get_location();
        if !location.is_undefined() {
            write!(f, " at: {location}")
        } else {
            Ok(())
        }
    }
}

impl Diagnostic {
    pub fn get_message(&self) -> &str {
        self.message.as_str()
    }

    pub fn get_location(&self) -> SourceLocation {
        self.primary_location.clone()
    }

    pub fn get_secondary_locations(&self) -> Option<&[SourceLocation]> {
        self.secondary_locations.as_deref()
    }

    pub fn get_error_code(&self) -> &'static str {
        self.error_code
    }

    pub fn get_sub_diagnostics(&self) -> &[Diagnostic] {
        &self.sub_diagnostics
    }
}

//Helper methods for diagnostics
impl Diagnostic {
    pub fn unexpected_token_found(expected: &str, found: &str, range: SourceLocation) -> Diagnostic {
        Diagnostic::new(format!("Unexpected token: expected {expected} but found {found}"))
            .with_error_code("E007")
            .with_location(range)
    }

    pub fn missing_function(location: SourceLocation) -> Diagnostic {
        Diagnostic::new("Cannot generate code outside of function context.")
            .with_error_code("E072")
            .with_location(location)
    }

    pub fn codegen_error(message: impl Into<String>, location: SourceLocation) -> Diagnostic {
        Diagnostic::new(message).with_location(location).with_error_code("E071")
    }

    pub fn llvm_error(file: &str, llvm_error: &str) -> Diagnostic {
        Diagnostic::new(format!("{file}: Internal llvm error: {:}", llvm_error)).with_error_code("E005")
    }

    pub fn missing_token(expected_token: &str, range: SourceLocation) -> Diagnostic {
        Diagnostic::new(format!("Missing expected Token {expected_token}"))
            .with_location(range)
            .with_error_code("E006")
    }

    pub fn invalid_parameter_count(expected: usize, received: usize, location: SourceLocation) -> Diagnostic {
        Diagnostic::new(
            format!(
                "Invalid parameter count. Received {received} parameters while {expected} parameters were expected.",
            )).with_error_code("E032")
            .with_location(location)
    }

    pub fn unknown_type(type_name: &str, location: SourceLocation) -> Diagnostic {
        Diagnostic::new(format!("Unknown type: {type_name:}")).with_error_code("E052").with_location(location)
    }

    pub fn unresolved_reference(reference: &str, location: SourceLocation) -> Diagnostic {
        Diagnostic::new(format!("Could not resolve reference to {reference:}"))
            .with_error_code("E048")
            .with_location(location)
    }

    pub fn invalid_assignment(right_type: &str, left_type: &str, location: SourceLocation) -> Diagnostic {
        Diagnostic::new(format!("Invalid assignment: cannot assign '{right_type}' to '{left_type}'"))
            .with_error_code("E037")
            .with_location(location)
    }

    pub fn cannot_generate_initializer(variable_name: &str, location: SourceLocation) -> Diagnostic {
        Self::new(format!(
            "Cannot generate literal initializer for '{variable_name}': Value cannot be derived"
        ))
        .with_error_code("E041")
        .with_location(location)
    }

    pub fn cannot_generate_call_statement(operator: &AstNode) -> Diagnostic {
        //TODO: We could probably get a better slice here
        Diagnostic::codegen_error(
            format!("cannot generate call statement for {:?}", operator),
            operator.get_location(),
        )
    }

    pub fn cannot_generate_from_empty_literal(type_name: &str, location: SourceLocation) -> Diagnostic {
        Diagnostic::codegen_error(
            format!("Cannot generate {type_name} from empty literal").as_str(),
            location,
        )
    }
}

// CFC related diagnostics
impl Diagnostic {
    pub fn unnamed_control(location: SourceLocation) -> Diagnostic {
        Diagnostic::new("Unnamed control").with_error_code("E087").with_location(location)
    }
}

#[cfg(test)]
mod tests {
    use codespan_reporting::files::{Location, SimpleFile};

    use crate::{diagnostics::Severity, reporter::clang::ClangFormatDiagnosticReporter};

    #[test]
    fn test_build_diagnostic_msg() {
        let reporter = ClangFormatDiagnosticReporter::default();
        let file = SimpleFile::new("test.st".to_string(), "source".to_string());
        let start = Location { line_number: 4, column_number: 1 };
        let end = Location { line_number: 4, column_number: 4 };
        let res = reporter.build_diagnostic_msg(
            Some(&file),
            Some(&start),
            Some(&end),
            "E001",
            &Severity::Error,
            "This is an error",
        );

        assert_eq!(res, "test.st:4:1:{4:1-4:4}: error[E001]: This is an error");
    }

    #[test]
    fn test_build_diagnostic_msg_equal_start_end() {
        let reporter = ClangFormatDiagnosticReporter::default();
        let file = SimpleFile::new("test.st".to_string(), "source".to_string());
        let start = Location { line_number: 4, column_number: 1 };
        let end = Location { line_number: 4, column_number: 1 };
        let res = reporter.build_diagnostic_msg(
            Some(&file),
            Some(&start),
            Some(&end),
            "E001",
            &Severity::Error,
            "This is an error",
        );

        assert_eq!(res, "test.st:4:1: error[E001]: This is an error");
    }

    #[test]
    fn test_build_diagnostic_msg_no_location() {
        let reporter = ClangFormatDiagnosticReporter::default();
        let file = SimpleFile::new("test.st".to_string(), "source".to_string());
        let res = reporter.build_diagnostic_msg(
            Some(&file),
            None,
            None,
            "E001",
            &Severity::Error,
            "This is an error",
        );

        assert_eq!(res, "test.st: error[E001]: This is an error");
    }

    #[test]
    fn test_build_diagnostic_msg_no_file() {
        let reporter = ClangFormatDiagnosticReporter::default();
        let start = Location { line_number: 4, column_number: 1 };
        let end = Location { line_number: 4, column_number: 4 };
        let res = reporter.build_diagnostic_msg(
            None,
            Some(&start),
            Some(&end),
            "E001",
            &Severity::Error,
            "This is an error",
        );

        assert_eq!(res, "error[E001]: This is an error");
    }

    #[test]
    fn test_build_diagnostic_msg_no_file_no_location() {
        let reporter = ClangFormatDiagnosticReporter::default();
        let res =
            reporter.build_diagnostic_msg(None, None, None, "E001", &Severity::Error, "This is an error");

        assert_eq!(res, "error[E001]: This is an error");
    }
}
