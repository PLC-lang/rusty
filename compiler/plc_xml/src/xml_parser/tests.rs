use ast::{
    ast::{
        flatten_expression_list, Assignment, AstNode, AstStatement, CallStatement, CompilationUnit,
        LinkageType,
    },
    provider::IdProvider,
};
use insta::assert_debug_snapshot;
use plc_diagnostics::diagnostics::Diagnostic;
use plc_source::SourceCode;

use crate::{
    serializer::{
        with_header, XBody, XConnection, XConnectionPointIn, XExpression, XFbd, XInVariable, XOutVariable,
        XPou, XRelPosition,
    },
    xml_parser::{self},
};

fn parse(content: &str) -> (CompilationUnit, Vec<Diagnostic>) {
    let source_code = SourceCode::new(content, "test.cfc");
    xml_parser::parse(&source_code, LinkageType::Internal, IdProvider::default())
}

#[test]
fn variable_assignment() {
    let pou = xml_parser::visit(content::ASSIGNMENT_A_B).unwrap();
    assert_debug_snapshot!(pou);
}

#[test]
fn conditional_return() {
    let statements = &parse(content::CONDITIONAL_RETURN).0.implementations[0].statements;
    assert_eq!(statements.len(), 2);
    assert_debug_snapshot!(statements[0]);
}

#[test]
fn conditional_return_negated() {
    let content =
        &content::CONDITIONAL_RETURN.replace(r#"<negated value="false"/>"#, r#"<negated value="true"/>"#);

    let statements = &parse(content).0.implementations[0].statements;

    assert_eq!(statements.len(), 2);
    assert_debug_snapshot!(statements[0]);
}

#[test]
fn conditional_return_without_connection() {
    let (_, diagnostics) = parse(content::CONDITIONAL_RETURN_WITHOUT_CONNECTION);
    assert_eq!(diagnostics.len(), 1);
    assert_debug_snapshot!(diagnostics);
}

#[test]
fn conditional_return_chained_to_another_conditional_return() {
    let (_, diagnostics) = parse(content::CONDITIONAL_RETURN_CHAINED_TO_ANOTHER_CONDITIONAL_RETURN);
    assert_eq!(diagnostics.len(), 2);
    assert_debug_snapshot!(diagnostics);
}

#[test]
fn model_is_sorted_by_execution_order() {
    let src = r#"
        <?xml version="1.0" encoding="UTF-8"?>
        <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="thistimereallyeasy" pouType="program">
            <interface>
                <localVars/>
                <addData>
                    <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                        <textDeclaration>
                            <content>
        PROGRAM thistimereallyeasy
        VAR
            a, b, c, d : DINT;
        END_VAR
                            </content>
                        </textDeclaration>
                    </data>
                </addData>
            </interface>
            <body>
                <FBD>
                    <inVariable localId="1" height="20" width="80" negated="false">
                        <position x="410" y="130"/>
                        <connectionPointOut>
                            <relPosition x="80" y="10"/>
                        </connectionPointOut>
                        <expression>a</expression>
                    </inVariable>
                    <outVariable localId="2" height="20" width="80" executionOrderId="2" negated="false" storage="none">
                        <position x="550" y="70"/>
                        <connectionPointIn>
                            <relPosition x="0" y="10"/>
                            <connection refLocalId="1"/>
                        </connectionPointIn>
                        <expression>b</expression>
                    </outVariable>
                    <outVariable localId="3" height="20" width="80" executionOrderId="0" negated="false" storage="none">
                        <position x="550" y="130"/>
                        <connectionPointIn>
                            <relPosition x="0" y="10"/>
                            <connection refLocalId="1"/>
                        </connectionPointIn>
                        <expression>c</expression>
                    </outVariable>
                    <outVariable localId="4" height="20" width="80" executionOrderId="1" negated="false" storage="none">
                        <position x="550" y="190"/>
                        <connectionPointIn>
                            <relPosition x="0" y="10"/>
                            <connection refLocalId="1"/>
                        </connectionPointIn>
                        <expression>d</expression>
                    </outVariable>
                </FBD>
            </body>
        </pou>
        "#;

    assert_debug_snapshot!(xml_parser::visit(src).unwrap());
}

#[test]
fn function_returns() {
    let content = with_header(
        &XPou::init(
            "FuncyReturn",
            "function",
            "FUNCTION FuncyReturn : DINT
                        VAR_INPUT
                            a : DINT;
                        END_VAR",
        )
        .with_body(
            XBody::new().with_fbd(
                XFbd::new()
                    .with_in_variable(
                        XInVariable::init("1", false).with_expression(XExpression::new().with_data("a")),
                    )
                    .with_out_variable(
                        XOutVariable::init("2", false)
                            .with_attribute("executionOrderId", "0")
                            .with_expression(XExpression::new().with_data("FuncyReturn"))
                            .with_connection_point_in(
                                XConnectionPointIn::new()
                                    .with_rel_position(XRelPosition::init().close())
                                    .with_connection(
                                        XConnection::new().with_attribute("refLocalId", "1").close(),
                                    ),
                            ),
                    ),
            ),
        )
        .serialize(),
    );

    assert_debug_snapshot!(xml_parser::visit(&content).unwrap());
}

#[test]
fn ast_generates_locations() {
    let source_code = SourceCode::new(content::CALL_BLOCK, "<internal>.cfc");
    let (units, diagnostics) = xml_parser::parse(&source_code, LinkageType::Internal, IdProvider::default());
    let impl1 = &units.implementations[0];
    //Deconstruct assignment and get locations
    let AstStatement::Assignment (Assignment{ left, right, .. })= &impl1.statements[0].get_stmt() else {
            panic!("Not an assignment");
        };
    assert_debug_snapshot!(left.get_location());
    assert_debug_snapshot!(right.get_location());
    //Deconstruct call statement and get locations
    let AstNode { stmt: AstStatement::CallStatement (CallStatement{ operator, parameters, .. }), location, ..} = &impl1.statements[1] else {
            panic!("Not a call statement");
        };
    assert_debug_snapshot!(location);
    assert_debug_snapshot!(operator.get_location());
    let parameters = parameters.as_deref().unwrap();
    let parameters = flatten_expression_list(parameters);
    for param in parameters {
        assert_debug_snapshot!(param.get_location());
    }

    assert_debug_snapshot!(impl1);
    assert!(diagnostics.is_empty());
}

#[test]
#[ignore = "Validation is not implemented on CFC tests yet, we need to be able to change parsers on the test utils level"]
fn ast_diagnostic_locations() {
    let source_code = SourceCode::new(content::ASSIGNMENT_TO_UNRESOLVED_REFERENCE, "<internal>.cfc");
    let (units, diagnostics) = xml_parser::parse(&source_code, LinkageType::Internal, IdProvider::default());
    let impl1 = &units.implementations[0];
    assert_debug_snapshot!(impl1);
    assert!(diagnostics.is_empty());
    //Run resolve and validate
    todo!("Validation in tests not yet done")
}

mod content {
    pub(super) const ASSIGNMENT_A_B: &str = r#"
        <?xml version="1.0" encoding="UTF-8"?>
        <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="thistimereallyeasy" pouType="program">
            <interface>
                <localVars/>
                <addData>
                    <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                        <textDeclaration>
                            <content>
        PROGRAM thistimereallyeasy
        VAR
            a, b : DINT;
        END_VAR
                            </content>
                        </textDeclaration>
                    </data>
                </addData>
            </interface>
            <body>
                <FBD>
                    <outVariable localId="2" height="20" width="80" executionOrderId="0" negated="false" storage="none">
                        <position x="550" y="130"/>
                        <connectionPointIn>
                            <relPosition x="0" y="10"/>
                            <connection refLocalId="1"/>
                        </connectionPointIn>
                        <expression>b</expression>
                    </outVariable>
                    <inVariable localId="1" height="20" width="80" negated="false">
                    <position x="410" y="130"/>
                    <connectionPointOut>
                        <relPosition x="80" y="10"/>
                    </connectionPointOut>
                    <expression>a</expression>
                </inVariable>
                </FBD>
            </body>
        </pou>
        "#;

    pub(super) const CALL_BLOCK: &str = r#"
        <?xml version="1.0" encoding="UTF-8"?>
        <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="program_0" pouType="program">
            <interface>
                <localVars/>
                <addData>
                    <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                        <textDeclaration>
                            <content>
        PROGRAM program_0
        VAR
        	x : DINT;
        	a : DINT;
        END_VAR
        					</content>
                        </textDeclaration>
                    </data>
                </addData>
            </interface>
            <body>
                <FBD>
                    <inVariable localId="1" height="20" width="80" negated="false">
                        <position x="280" y="80"/>
                        <connectionPointOut>
                            <relPosition x="80" y="10"/>
                        </connectionPointOut>
                        <expression>x</expression>
                    </inVariable>
                    <outVariable localId="2" height="20" width="80" executionOrderId="0" negated="false" storage="none">
                        <position x="520" y="170"/>
                        <connectionPointIn>
                            <relPosition x="0" y="10"/>
                            <connection refLocalId="1">
                                <position x="520" y="180"/>
                                <position x="430" y="180"/>
                                <position x="430" y="90"/>
                                <position x="360" y="90"/>
                            </connection>
                        </connectionPointIn>
                        <expression>a</expression>
                    </outVariable>
                    <block localId="3" width="60" height="60" typeName="ADD" executionOrderId="1">
                        <position x="190" y="160"/>
                        <inputVariables>
                            <variable formalParameter="" negated="false">
                                <connectionPointIn>
                                    <relPosition x="0" y="30"/>
                                    <connection refLocalId="4"/>
                                </connectionPointIn>
                            </variable>
                            <variable formalParameter="" negated="false">
                                <connectionPointIn>
                                    <relPosition x="0" y="50"/>
                                    <connection refLocalId="5"/>
                                </connectionPointIn>
                            </variable>
                        </inputVariables>
                        <inOutVariables/>
                        <outputVariables>
                            <variable formalParameter="" negated="false">
                                <connectionPointOut>
                                    <relPosition x="60" y="30"/>
                                </connectionPointOut>
                            </variable>
                        </outputVariables>
                    </block>
                    <inVariable localId="4" height="20" width="80" negated="false">
                        <position x="40" y="180"/>
                        <connectionPointOut>
                            <relPosition x="80" y="10"/>
                        </connectionPointOut>
                        <expression>a</expression>
                    </inVariable>
                    <inVariable localId="5" height="20" width="80" negated="false">
                        <position x="40" y="200"/>
                        <connectionPointOut>
                            <relPosition x="80" y="10"/>
                        </connectionPointOut>
                        <expression>1</expression>
                    </inVariable>
                </FBD>
            </body>
        </pou>
        
    "#;

    pub(super) const ASSIGNMENT_TO_UNRESOLVED_REFERENCE: &str = r#"
        <?xml version="1.0" encoding="UTF-8"?>
        <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="program_0" pouType="program">
            <interface>
                <localVars/>
                <addData>
                    <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                        <textDeclaration>
                            <content>
        PROGRAM program_0
        VAR
        	x : DINT;
        END_VAR
        					</content>
                        </textDeclaration>
                    </data>
                </addData>
            </interface>
            <body>
                <FBD>
                    <inVariable localId="1" height="20" width="80" negated="false">
                        <position x="280" y="80"/>
                        <connectionPointOut>
                            <relPosition x="80" y="10"/>
                        </connectionPointOut>
                        <expression>x</expression>
                    </inVariable>
                    <outVariable localId="2" height="20" width="80" executionOrderId="0" negated="false" storage="none">
                        <position x="520" y="170"/>
                        <connectionPointIn>
                            <relPosition x="0" y="10"/>
                            <connection refLocalId="1">
                                <position x="520" y="180"/>
                                <position x="430" y="180"/>
                                <position x="430" y="90"/>
                                <position x="360" y="90"/>
                            </connection>
                        </connectionPointIn>
                        <expression>a</expression>
                    </outVariable>
                </FBD>
            </body>
        </pou>
        
    "#;

    pub(super) const CONDITIONAL_RETURN: &str = r#"
    <?xml version="1.0" encoding="UTF-8"?>
    <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="conditional_return" pouType="functionBlock">
        <interface>
            <localVars/>
            <addData>
                <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                    <textDeclaration>
                        <content>
    FUNCTION_BLOCK conditional_return
    VAR_INPUT
        val : DINT;
    END_VAR</content>
                    </textDeclaration>
                </data>
            </addData>
        </interface>
        <body>
            <FBD>
                <inVariable localId="1" height="20" width="82" negated="false">
                    <position x="220" y="60"/>
                    <connectionPointOut>
                        <relPosition x="82" y="10"/>
                    </connectionPointOut>
                    <expression>val = 5</expression>
                </inVariable>
                <return localId="2" height="20" width="76" executionOrderId="0">
                    <position x="330" y="60"/>
                    <connectionPointIn>
                        <relPosition x="0" y="10"/>
                        <connection refLocalId="1"/>
                    </connectionPointIn>
                    <addData>
                        <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                            <negated value="false"/>
                        </data>
                    </addData>
                </return>
                <inVariable localId="3" height="20" width="80" negated="false">
                    <position x="220" y="100"/>
                    <connectionPointOut>
                        <relPosition x="80" y="10"/>
                    </connectionPointOut>
                    <expression>10</expression>
                </inVariable>
                <outVariable localId="4" height="20" width="80" executionOrderId="1" negated="false" storage="none">
                    <position x="330" y="100"/>
                    <connectionPointIn>
                        <relPosition x="0" y="10"/>
                        <connection refLocalId="3"/>
                    </connectionPointIn>
                    <expression>val</expression>
                </outVariable>
                <inOutVariable localId="5" height="20" width="80" negatedIn="false" storageIn="none" negatedOut="false">
                    <position x="780" y="60"/>
                    <connectionPointIn>
                        <relPosition x="0" y="10"/>
                    </connectionPointIn>
                    <connectionPointOut>
                        <relPosition x="80" y="10"/>
                    </connectionPointOut>
                    <expression>a</expression>
                </inOutVariable>
            </FBD>
        </body>
    </pou>
    "#;

    pub(super) const CONDITIONAL_RETURN_WITHOUT_CONNECTION: &str = r#"
    <?xml version="1.0" encoding="UTF-8"?>
    <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="conditional_return" pouType="functionBlock">
        <interface>
            <localVars/>
            <addData>
                <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                    <textDeclaration>
                        <content>
    FUNCTION_BLOCK conditional_return
    VAR_INPUT
        val : DINT;
    END_VAR</content>
                    </textDeclaration>
                </data>
            </addData>
        </interface>
        <body>
            <FBD>
                <inVariable localId="1" height="20" width="82" negated="false">
                    <position x="220" y="60"/>
                    <connectionPointOut>
                        <relPosition x="82" y="10"/>
                    </connectionPointOut>
                    <expression>val = 5</expression>
                </inVariable>
                <return localId="2" height="20" width="76" executionOrderId="0">
                    <position x="330" y="60"/>
                    <addData>
                        <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                            <negated value="false"/>
                        </data>
                    </addData>
                </return>
                <inVariable localId="3" height="20" width="80" negated="false">
                    <position x="220" y="100"/>
                    <connectionPointOut>
                        <relPosition x="80" y="10"/>
                    </connectionPointOut>
                    <expression>10</expression>
                </inVariable>
                <outVariable localId="4" height="20" width="80" executionOrderId="1" negated="false" storage="none">
                    <position x="330" y="100"/>
                    <connectionPointIn>
                        <relPosition x="0" y="10"/>
                        <connection refLocalId="3"/>
                    </connectionPointIn>
                    <expression>val</expression>
                </outVariable>
                <inOutVariable localId="5" height="20" width="80" negatedIn="false" storageIn="none" negatedOut="false">
                    <position x="780" y="60"/>
                    <connectionPointIn>
                        <relPosition x="0" y="10"/>
                    </connectionPointIn>
                    <connectionPointOut>
                        <relPosition x="80" y="10"/>
                    </connectionPointOut>
                    <expression>a</expression>
                </inOutVariable>
            </FBD>
        </body>
    </pou>
    "#;

    pub(super) const CONDITIONAL_RETURN_CHAINED_TO_ANOTHER_CONDITIONAL_RETURN: &str = r#"
    <?xml version="1.0" encoding="UTF-8"?>
    <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="conditional_return" pouType="functionBlock">
        <interface>
            <localVars/>
            <addData>
                <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                    <textDeclaration>
                        <content>
    FUNCTION_BLOCK conditional_return
    VAR_INPUT
        val : DINT;
    END_VAR</content>
                    </textDeclaration>
                </data>
            </addData>
        </interface>
        <body>
            <FBD>
                <return localId="1" height="20" width="76" executionOrderId="0">
                    <position x="330" y="60"/>
                    <addData>
                        <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                            <negated value="false"/>
                        </data>
                    </addData>
                </return>
                <return localId="2" height="20" width="76" executionOrderId="1">
                    <position x="330" y="60"/>
                    <connectionPointIn>
                        <relPosition x="0" y="10"/>
                        <connection refLocalId="1"/>
                    </connectionPointIn>
                    <addData>
                        <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                            <negated value="false"/>
                        </data>
                    </addData>
                </return>
            </FBD>
        </body>
    </pou>
    "#;
}
