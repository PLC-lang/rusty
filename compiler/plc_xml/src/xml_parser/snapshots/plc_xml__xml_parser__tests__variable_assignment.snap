---
source: compiler/plc_xml/src/xml_parser/tests.rs
expression: pou
---
Project {
    pous: [
        Pou {
            name: "thistimereallyeasy",
            pou_type: Program,
            body: Body {
                function_block_diagram: FunctionBlockDiagram {
                    nodes: {
                        1: FunctionBlockVariable(
                            FunctionBlockVariable {
                                kind: Input,
                                local_id: 1,
                                negated: false,
                                expression: "a",
                                execution_order_id: None,
                                ref_local_id: None,
                            },
                        ),
                        2: FunctionBlockVariable(
                            FunctionBlockVariable {
                                kind: Output,
                                local_id: 2,
                                negated: false,
                                expression: "b",
                                execution_order_id: Some(
                                    0,
                                ),
                                ref_local_id: Some(
                                    1,
                                ),
                            },
                        ),
                    },
                },
            },
            actions: [],
            interface: Some(
                Interface {
                    add_data: Some(
                        Data {
                            content: "PROGRAM thistimereallyeasy\n        VAR\n            a, b : DINT;\n        END_VAR\nEND_PROGRAM",
                            handle: Implementation,
                        },
                    ),
                },
            ),
        },
    ],
}
